src/theory/theory_model.cpp

```diff
@@ -803,15 +804,57 @@ std::string TheoryModel::debugPrintModelEqc() const
   return ss.str();
 }
 
-bool TheoryModel::isValue(TNode node)
+struct IsModelValueTag
 {
-  if (node.isConst())
+};
+struct IsModelValueComputedTag
+{
+};
+/** Attribute true for expressions that are quasi-values */
+using IsModelValueAttr = expr::Attribute<IsModelValueTag, bool>;
+using IsModelValueComputedAttr = expr::Attribute<IsModelValueComputedTag, bool>;
+
+bool TheoryModel::isValue(TNode n) const
+{
+  if (!n.getAttribute(IsModelValueComputedAttr()))
   {
-    return true;
+    bool isQv = false;
+    if (n.isConst())
+    {
+      isQv = true;
+    }
+    else
+    {
+      Kind k = n.getKind();
+      if (k == kind::REAL_ALGEBRAIC_NUMBER || k == kind::LAMBDA
+          || k == kind::WITNESS)
+      {
+        // we are a value if we are one of the above kinds
+        isQv = true;
+      }
+      else if (n.getNumChildren() > 0 && rewrite(n) == n)
+      {
+        // note that we must be in rewritten form
+        isQv = true;
+        for (TNode nc : n)
+        {
+          if (!isValue(nc))
```

Comment:
It would be nice if we could make this method iterative instead of recursive. The terms passed to this function are probably not deeply nested in most cases, but it'd be nice for consistency.

Reply:
I've done this, although notice that the algorithm is really complex now to account for how the original computation worked.